### PR TITLE
Validate trusted_cluster name

### LINKED ISSUE #48309: `trusted_cluster` resource renames itself, breaking subsequent updates
### Expected behaviour:

- As a user I create a resource called `clusterA` while my remote cluster real name is `clusterA.example.com`
Either I have a name mismatch error when I create the resource, or the resource name stays `clusterA`

### Current behaviour:

- The resource is renamed from `clusterA` to `clusterA.example.com` in the backend
- I don't have a unique identifier to get the resource anymore
- If I edit the role mapping and apply again for `clusterA` it silently drop my changes

### Bug details:
- Teleport version v16.4.2
- Recreation steps: create a TC resource whose name is different than the root cluster name
- Debug logs

### Additional notes

- this is blocking TC support in the operator

### PR DESCRIPTION
**Pull request by bernardjkim**
Closes https://github.com/gravitational/teleport/issues/48309

Teleport now prevents the creation of a `trusted_cluster` resource if the name does not match the cluster name.  Users will be shown the following error message if they attempt to create a `teleport_cluster` with an invalid cluster name:
> trusted cluster resource name must be the same as the remote cluster name. got: "root", expected: "root.example.com"

changelog: The trusted_cluster resource name must now match the cluster name on creation

### CODE REVIEW

**Comment by hugoShaka** [comment]
This is a strongly breaking change as we had no such limitation in the past. Most of our users TC resources are created with names not matching the trusted cluster name. I don't think we can pull this one, even in a major version. This will cause a lot of toil on both users and customer support.

Have you considered other alternatives that will not break everyone's IaC?

File: lib/auth/trustedcluster.go [hunk 1]
```diff
@@ -322,9 +322,10 @@ func (a *Server) establishTrust(ctx context.Context, trustedCluster types.Truste
 			if remoteClusterName == domainName {
 				return nil, trace.BadParameter("remote cluster name can not be the same as local cluster name")
 			}
-			// TODO(klizhentas) in 2.5.0 prohibit adding trusted cluster resource name
-			// different from cluster name (we had no way of checking this before x509,
-			// because SSH CA was a public key, not a cert with metadata)
```
Comment (hugoShaka) [inline lib/auth/trustedcluster.go:327]: Also, gg for addressing a TODO from Sasha in v2 😅

**Comment by hugoShaka** [comment]
paging @espadolini for the trusted cluster expertise 
historically accurate picture of Edoardo when I'm asking for help on Slack

**Comment by bernardjkim** [comment]
Validating the name seemed like the simplest approach to addressing the issue, but breaking existing automation is definitely the big concern. I'll explore some alternate solutions.

**Comment by hugoShaka** [comment]
> but breaking existing automation is definitely the big concern. I'll explore some alternate solutions.

If we introduce a new resource version and run the validation only if `tc.version == types.V3` it should not break everyone's automation and allow us to build sane integrations.

I explored storing both the resource name and the remote cluster name in different fields, but this might also pose challenges:
- inconsistent behaviours when trying to apply resources we already renamed before doing this change
- might face indexing issues as we try to lookup resources with their name or trusted_cluster_name depending on the context

**Comment by espadolini** [comment]
Is there any automation that doesn't break horribly when a created resource gets renamed under its nose?

**Comment by hugoShaka** [comment]
> Is there any automation that doesn't break horribly when a created resource gets renamed under its nose?

Sadly, terraform can import back state. Also if you just `tctl create -f` your whole directory, you might not notice the renaming madness. But yes, any decent automation would explode when losing its resource unique identifier. That's why we must fix this before adding operator support.

**Comment by espadolini** [comment]
I wouldn't be against just breaking compatibility for this, even without a (meaningless) version bump - but we'd only get to do that on a major version change, so it'd have to be v18-only - are we ok with that?

**Comment by Tener** [comment]
What is the rationale behind the renaming anyway? Maybe we can/should drop _that_ behaviour instead?

**Comment by hugoShaka** [comment]
> even without a (meaningless) resource version bump

I disagree on not bumping the resource version: resource versions are not only here to tell us how the manifest should be parsed. We also use them to protect users against breaking changes and ensure deterministic behaviour when applying manifests.

We've been doing this semi-successfully for roles to avoid breaking existing manifests. Adding a new resource version costs us close to nothing and makes the breaking change visible to the user. This allows us to backport the new behaviour to any supported Teleport version, and we can still deprecate and remove the old renaming version in v18 or v19 if we want.

> What is the rationale behind the renaming anyway? Maybe we can/should drop that behaviour instead?

Naming TC resources after the TC cluster name allows us to lookup the TC resource by:
- looking up at the SAN of who signed the certificate
- doing a single backend lookup to get the TC resource
- checking if the cert was signed by the trusted cluster's CA.

Not naming the TC resource like the cluster name would require us to iterate over every trusted cluster each time to know if a cert is trusted (or build an in-memory map) and keep track of the cluster name anyway to know which CA might haver signed this.

Also, if we were to remove the renaming logic, users who wrote TC manifests whose name don't match the cluster name will create new TCs when applying again their manifest, instead of editing the current one.

**Comment by bernardjkim** [comment]
Superseded by https://github.com/gravitational/teleport/pull/49789

**Pull request closed without merging**